filament/src/details/InstanceBuffer.cpp

```diff
@@ -0,0 +1,127 @@
+}
+
+BuilderType::Builder::~Builder() noexcept = default;
+BuilderType::Builder::Builder(BuilderType::Builder const& rhs) noexcept = default;
+BuilderType::Builder::Builder(BuilderType::Builder&& rhs) noexcept = default;
+BuilderType::Builder& BuilderType::Builder::operator=(BuilderType::Builder const& rhs) noexcept = default;
+BuilderType::Builder& BuilderType::Builder::operator=(BuilderType::Builder&& rhs) noexcept = default;
+
+InstanceBuffer::Builder& InstanceBuffer::Builder::localTransforms(
+        math::mat4f const* localTransforms) {
+    mImpl->mLocalTransforms = localTransforms;
+    return *this;
+}
+
+InstanceBuffer* InstanceBuffer::Builder::build(Engine& engine) {
+    return downcast(engine).createInstanceBuffer(*this);
+}
+
+// ------------------------------------------------------------------------------------------------
+
+FInstanceBuffer::FInstanceBuffer(FEngine& engine, const Builder& builder) {
+    mInstanceCount = builder->mInstanceCount;
+
+    mLocalTransforms.reserve(mInstanceCount);
+    mLocalTransforms.resize(mInstanceCount);    // this will initialize all transforms to identity
+
+    if (builder->mLocalTransforms) {
+        memcpy(mLocalTransforms.data(), builder->mLocalTransforms,
+                sizeof(math::mat4f) * mInstanceCount);
+    }
+
+    // Allocate our instance buffer. We always allocate a size to match PerRenderableUib, regardless
+    // of the number of instances. This is because the buffer will get bound to the PER_RENDERABLE
+    // UBO, and we can't bind a buffer smaller than the full size of the UBO.
+    DriverApi& driver = engine.getDriverApi();
+    mUboHandle = driver.createBufferObject(sizeof(PerRenderableUib),
+            BufferObjectBinding::UNIFORM, backend::BufferUsage::STATIC);
+}
+
+void FInstanceBuffer::setLocalTransforms(
+        math::mat4f const* localTransforms, size_t count, size_t offset) {
+    ASSERT_PRECONDITION(offset + count <= mInstanceCount,
+            "setLocalTransforms overflow. InstanceBuffer has only %zu instances, but trying to set "
+            "%zu transforms at offset %zu.",
+            mInstanceCount,
+            count,
+            offset);
+    memcpy(mLocalTransforms.data() + offset, localTransforms, sizeof(math::mat4f) * count);
+}
+
+void FInstanceBuffer::prepare(FEngine& engine, math::mat4f rootTransform, PerRenderableData& ubo) {
+    DriverApi& driver = engine.getDriverApi();
+
+    // TODO: keep track of dirtiness. If the rootTransform and localTransforms have not changed,
+    // there's no need to update the buffer.
+
+    // TODO: allocate this staging buffer from a pool.
+    uint32_t stagingBufferSize = sizeof(PerRenderableUib);
+    PerRenderableData* stagingBuffer = (PerRenderableData*)::malloc(stagingBufferSize);
+    // TODO: consider using JobSystem to parallelize this.
+    for (size_t i = 0; i < mInstanceCount; i++) {
+        stagingBuffer[i] = ubo;
+        math::mat4f model = rootTransform * mLocalTransforms[i];
```

Comment:
Are you sure it's not `mLocalTransforms[i] * rootTransform` ? Shouldn't we apply the rootTransform first, and then the local transform?

Reply:
I'm fairly certain.

Say the user wants to create a forrest of instanced trees, each with a random translation and scale. Each tree’s “local transform” could be computed like:

```
mat4f localTransform =
    mat4f::translation(float3{rand(), 0.0f, rand()}) *          // call this T
    mat4f::rotation(rand(), float3{0.0f, 1.0f, 0.0f});          // call this R
```

To move the forest as a whole, the user can simply adjust the rootTransform (renderable's transform). Assume they set it to a translation T'.

If the computation was `mLocalTransforms[i] * rootTransform`, then we'd end up with:

```
worldSpace = T * R * T' * vertex
```

That isn't correct, or at least very unintuitive. This is also complicated by the fact that T' can contain the world offset transformation.

Instead, we want `rootTransform * mLocalTransforms[i]`:

```
worldSpace = T' * T * R * vertex
```

Reply:
you're right, I got myself confused.